### PR TITLE
monasca: Add Apache Cassandra support (SCRD-6731)

### PR DESCRIPTION
**Pull request by witekest**
The change adds the option to install Apache Cassandra as the TSDB.
Currently only single node installation is supported.
Choosing this option implicitly installs Java implementation of
monasca-persister.

### CODE REVIEW

File: chef/data_bags/crowbar/template-monasca.json [hunk 1]
```diff
@@ -124,15 +124,16 @@
...
+      },
       "master": {
         "influxdb_retention_policy": "60d",
-        "cassandra_admin_password": "",
         "database_notification_password": "",
         "database_logapi_password": "",
         "keystone_monasca_operator_password": "",
         "keystone_monasca_agent_password": "",
         "keystone_admin_agent_password": "",
-        "tsdb": "influxdb",
```
Comment (jgrassler) [inline chef/data_bags/crowbar/template-monasca.json]: Just a question: why did you move this away from master? I'm ok with it, but it _is_ going to require a documentation update :-)

File: chef/cookbooks/monasca/recipes/cassandra.rb [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+package "cassandra"
+
+monasca_servers = search(:node, "roles:monasca-server")
```
Comment (toabctl) [inline chef/cookbooks/monasca/recipes/cassandra.rb]: there is `node_search_with_cache` . any reason why tho not use that? In envs with a lot of nodes, `node_search_with_cache` is prefered.

File: chef/cookbooks/monasca/attributes/default.rb [hunk 3]
```diff
@@ -108,6 +108,18 @@
...
 default[:monasca][:influxdb][:client_port] = 8086
 
+# cassandra
+default[:monasca][:cassandra][:user] = "cassandra"
+default[:monasca][:cassandra][:group] = "cassandra"
+default[:monasca][:cassandra][:client_port] = 9042
+default[:monasca][:cassandra][:batch_size_warn_threshold_in_kb] = 2000
+default[:monasca][:cassandra][:batch_size_fail_threshold_in_kb] = 2500
+default[:monasca][:cassandra][:admin_user] = "cassandra"
+default[:monasca][:cassandra][:admin_default_password] = "cassandra"
```
Comment (toabctl) [inline chef/cookbooks/monasca/attributes/default.rb]: is that the password that is used when the package is freshly installed?

File: chef/cookbooks/monasca/recipes/cassandra.rb [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+persister_role = node[:monasca][:cassandra][:monasca_persister_role]
+default_password = node[:monasca][:cassandra][:admin_default_password]
+admin_password = node[:monasca][:cassandra][:admin_password]
+cassandra_host = monasca_monitoring_host
+
+ruby_block "Initialize password for superuser role" do
+  block do
+    CassandraHelper.set_password(admin_role, admin_password,
+                                 user: admin_user,
+                                 password: default_password,
```
Comment (toabctl) [inline chef/cookbooks/monasca/recipes/cassandra.rb]: shouldn't that be the `admin_password` ?

File: chef/cookbooks/monasca/recipes/monasca_api.rb [hunk 5]
```diff
@@ -68,7 +70,7 @@
     node.set[:monasca][:db_monapi_synced] = true
     node.save
   end
-  not_if { node[:monasca][:db_monapi_synced] }
+  not_if {node[:monasca][:db_monapi_synced]}
```
Comment (toabctl) [inline chef/cookbooks/monasca/recipes/monasca_api.rb]: change not needed

File: chef/cookbooks/monasca/recipes/monasca_persister_java.rb [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+package "openstack-monasca-persister-java"
+
+monasca_servers = search(:node, "roles:monasca-server")
```
Comment (toabctl) [inline chef/cookbooks/monasca/recipes/monasca_persister_java.rb]: `node_search_with_cache` imo

File: chef/cookbooks/monasca/recipes/monasca_persister_java.rb [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+monasca_net_ip = MonascaHelper.get_host_for_monitoring_url(monasca_servers[0])
+
+template "/etc/monasca-persister/monasca-persister.yaml" do
+  source "monasca-persister-java.yaml.erb"
+  owner node[:monasca][:persister][:user]
+  group node[:monasca][:persister][:group]
+  mode "0640"
+  variables(
+    zookeeper_host: monasca_net_ip,
```
Comment (toabctl) [inline chef/cookbooks/monasca/recipes/monasca_persister_java.rb]: where is `zookeeper_host` defined?

File: chef/cookbooks/monasca/recipes/monasca_persister_java.rb [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+    cassandra_hosts: monasca_net_ip
```
Comment (toabctl) [inline chef/cookbooks/monasca/recipes/monasca_persister_java.rb]: same here

File: chef/data_bags/crowbar/template-monasca.json (hunk 1 above)
Comment (toabctl) [reply chef/data_bags/crowbar/template-monasca.json]: There is no `master` role anymore. I think we should try to remove all remaining options from master

Comment (witekest) [reply chef/data_bags/crowbar/template-monasca.json]: +1

File: chef/cookbooks/monasca/recipes/cassandra.rb (hunk 2 above)
Comment (witekest) [reply chef/cookbooks/monasca/recipes/cassandra.rb]: I copied it from your code in InfluxDB recipe :)
The point is valid and I fixed it in both recipes.

File: chef/cookbooks/monasca/attributes/default.rb (hunk 3 above)
Comment (witekest) [reply chef/cookbooks/monasca/attributes/default.rb]: Yes.

File: chef/cookbooks/monasca/recipes/cassandra.rb (hunk 4 above)
Comment (witekest) [reply chef/cookbooks/monasca/recipes/cassandra.rb]: No, we log in with `default_password` and set it to `admin_password`.

File: chef/cookbooks/monasca/recipes/monasca_api.rb (hunk 5 above)
Comment (witekest) [reply chef/cookbooks/monasca/recipes/monasca_api.rb]: Reverted.

File: chef/cookbooks/monasca/recipes/monasca_persister_java.rb (hunk 6 above)
Comment (witekest) [reply chef/cookbooks/monasca/recipes/monasca_persister_java.rb]: Fixed both in _monasca_perister_ and _monasca_persister_java_ recipes.

File: chef/cookbooks/monasca/recipes/monasca_persister_java.rb (hunk 7 above)
Comment (witekest) [reply chef/cookbooks/monasca/recipes/monasca_persister_java.rb]: I don't understand the question. `zookeeper_host` is the variable used in _monasca-persister-java.yaml.erb_ template.

Comment (toabctl) [reply chef/cookbooks/monasca/recipes/monasca_persister_java.rb]: urgs. not enough tee in the morning. sorry

File: chef/data_bags/crowbar/template-monasca.json (hunk 1 above)
Comment (witekest) [reply chef/data_bags/crowbar/template-monasca.json]: > Just a question: why did you move this away from master? I'm ok with it, but it _is_ going to require a documentation update :-)

Setting passwords is [not documented](http://docserv.nue.suse.com/documents/SUSE_OpenStack_Cloud_9/suse-openstack-cloud-deployment/single-html/#id-1.3.5.4.25.18).

**Comment by toabctl** [comment]
@witekest I tested this manually and it failed with:
```
FATAL: RuntimeError: ruby_block[Create role for monasca_api] (monasca::cassandra line 68) had an error: RuntimeError: Can not execute '/usr/bin/cqlsh --user cassandra --password gluCm3hXXwxp --execute "CREATE ROLE IF NOT EXISTS monasca_api WITH PASSWORD='gluCm3hXXwxp' AND LOGIN = true" 192.168.192.82': Connection error: ('Unable to connect to any servers', {'192.168.192.82': error(111, "Tried connecting to [('192.168.192.82', 9042)]. Last error: Connection refused")})
```

**Comment by witekest** [comment]
> @witekest I tested this manually and it failed with:
> 
> ```
> FATAL: RuntimeError: ruby_block[Create role for monasca_api] (monasca::cassandra line 68) had an error: RuntimeError: Can not execute '/usr/bin/cqlsh --user cassandra --password gluCm3hXXwxp --execute "CREATE ROLE IF NOT EXISTS monasca_api WITH PASSWORD='gluCm3hXXwxp' AND LOGIN = true" 192.168.192.82': Connection error: ('Unable to connect to any servers', {'192.168.192.82': error(111, "Tried connecting to [('192.168.192.82', 9042)]. Last error: Connection refused")})
> ```
> 
> You might need to wait a bit or retry after cassandra starts. There is no SD_NOTIFY support in cassandra

Thanks for catching. I have moved this section to _monasca-api_ recipe, so Cassandra should have enough time to start. I have also added `retries` in the following block, as I got AuthenticationError when testing in the same environment.

**Comment by toabctl** [comment]
Triggered a CI build with monasca/cassandra
